### PR TITLE
Add toByteBuffer to BinaryData

### LINKED ISSUE #19665: Add support to get Flux<ByteBuffer> from BinaryData
Currently, BinaryData has support for creating an instance using `fromFlux` and there's no corresponding getter method to get ByteBuffer flux. So, the user has to do this `Flux.just(ByteBuffer.wrap(data.toBytes()))` instead of just `data.toFlux()` or `data.toByteBufferFlux()`. Another issue is that `data.toBytes()` will copy the array and can be avoided by adding this API.

See this comment for more context: https://github.com/Azure/azure-sdk-for-java/pull/19607#discussion_r588594460

### PR DESCRIPTION
**Pull request by alzimmermsft**
Fixes #19665 

This PR adds a new API to `BinaryData` called `toByteBuffer`. The new API returns a read-only `ByteBuffer` view of the underlying `BinaryData`. This change allows for performance optimizations when using a `BinaryData`, or many `BinaryData`s, in a `Flux<ByteBuffer>`, previously this required buffering the data again as `BinaryData.toBytes` performs a copy of the underlying data.

The API change here was chosen instead of offer `toFluxByteBuffer` as this is more generic and can be used by a `Flux<ByteBuffer>` while also still supporting other APIs which use `ByteBuffer` directly.

cc: @kasobol-msft

### CODE REVIEW

File: sdk/core/azure-core/src/main/java/com/azure/core/util/BinaryData.java [hunk 1]
```diff
@@ -593,6 +594,21 @@ public InputStream toStream() {
...
+     * <p>
+     * Attempting to mutate the returned {@link ByteBuffer} will throw a {@link ReadOnlyBufferException}.
+     *
+     * <p><strong>Get a read-only ByteBuffer from the BinaryData</strong></p>
+     *
+     * {@codesnippet com.azure.util.BinaryData.toReadOnlyByteBuffer}
+     *
+     * @return A read-only {@link ByteBuffer} representing the {@link BinaryData}.
+     */
+    public ByteBuffer toReadOnlyByteBuffer() {
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/BinaryData.java:608]: Do we want to shorten this to `toByteBuffer` and just document that it is read-only?

@srnagar, @JonathanGiles

Comment (JonathanGiles) [reply sdk/core/azure-core/src/main/java/com/azure/core/util/BinaryData.java:608]: I like that suggestion

Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/util/BinaryData.java]: Changed to `toByteBuffer` with documentation explaining that it is read-only.

File: sdk/core/azure-core/src/main/java/com/azure/core/util/BinaryData.java [hunk 2]
```diff
@@ -593,6 +594,21 @@ public InputStream toStream() {
         return new ByteArrayInputStream(this.data);
     }
 
+    /**
+     * Returns a read-only {@link ByteBuffer} representation of this {@link BinaryData}.
+     * <p>
+     * Attempting to mutate the returned {@link ByteBuffer} will throw a {@link ReadOnlyBufferException}.
```
Comment (JonathanGiles) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/BinaryData.java:600]: One day our lack of proper HTML5 quality JavaDoc will come back to bite us hard. We need to have matching opening / closing elements.

Comment (kasobol-msft) [reply sdk/core/azure-core/src/main/java/com/azure/core/util/BinaryData.java:600]: The codebase around convinced me this is the way... Is there any checkstyle rule that would enforce this?

**Review by kasobol-msft — APPROVED** [review]

File: sdk/core/azure-core/src/main/java/com/azure/core/util/BinaryData.java [hunk 3]
```diff
@@ -593,6 +594,21 @@ public InputStream toStream() {
...
+     *
+     * <p><strong>Get a read-only ByteBuffer from the BinaryData</strong></p>
+     *
+     * {@codesnippet com.azure.util.BinaryData.toByteBuffer}
+     *
+     * @return A read-only {@link ByteBuffer} representing the {@link BinaryData}.
+     */
+    public ByteBuffer toByteBuffer() {
+        return ByteBuffer.wrap(this.data).asReadOnlyBuffer();
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/BinaryData.java:609]: Can this be done just once and cache the ByteBuffer instead of wrapping the data each time?

**Review by srnagar — APPROVED** [review]

File: sdk/core/azure-core/src/main/java/com/azure/core/util/BinaryData.java (hunk 3 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/util/BinaryData.java:609]: It could be cached but each return would need to be a `duplicate` (another `ByteBuffer` view on top of the current one) and that would effectively be the same cost as this.
